File: 2023/09-mirage-maintenance/main.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"slices"
	"strings"
)

func predict(a []int) int {
	if len(a) == 0 || AllEqual(a, 0) {
		return 0
	}

	nexta := []int{}
	for i := 0; i < len(a)-1; i++ {
		nexta = append(nexta, a[i+1]-a[i])
	}

	return a[len(a)-1] + predict(nexta)
}

func solve(f string, reverse bool) (sum int) {
	in, _ := os.ReadFile(f)

	for _, l := range strings.Split(strings.TrimSpace(string(in)), "\n") {
		a := []int{}
		json.Unmarshal([]byte("["+strings.Join(strings.Fields(l), ",")+"]"), &a)
		if reverse {
			slices.Reverse(a)
		}
		sum += predict(a)
	}
	return
}

func main() {
	fmt.Println("Day 9: Mirage Maintenance")
	fmt.Println("\tPart One:", solve("../aoc-inputs/2023/09/input.txt", false)) // 1921197370
	fmt.Println("\tPart Two:", solve("../aoc-inputs/2023/09/input.txt", true))  // 1124
}

func AllEqual(a []int, s int) bool {
	for i := range a {
		if a[i] != s {
			return false
		}
	}
	return true
}

File: 2023/09-mirage-maintenance/main.js
const fs = require('node:fs');

const data = fs.readFileSync('../aoc-inputs/2023/09/input.txt', 'utf8');

let predict = function(a) {
	if (a.every(function(v){return v==0})) {
		return 0
	}

	next = new Array(a.length-1)
	for (let i=0; i<a.length-1; i++) {
		next[i] = a[i+1]-a[i]
	}

	return a[a.length-1] + predict(next)
}

console.log("Day 9: Mirage Maintenance")

const numbers = data.split('\n');

let next = 0
next = numbers.reduce(
	(acc, curr) => acc + predict(curr.split(' ').map(Number)), next
)
console.log("\tPart One:",next) // 1921197370

let prev = 0
prev = numbers.reduce(
	(acc, curr) => acc + predict(curr.split(' ').map(Number).reverse()), prev
)
console.log("\tPart Two:",prev) // 1124